### PR TITLE
rpcdaemon: use BlockNum instead of uint64_t as block number type

### PR DESCRIPTION
**Pull request by lupin012**

### CODE REVIEW

File: silkworm/silkrpc/commands/debug_api.cpp [hunk 1]
```diff
@@ -625,7 +625,7 @@ Task<void> DebugRpcApi::handle_debug_trace_block_by_hash(const nlohmann::json& r
     co_return;
 }
 
-Task<std::set<evmc::address>> get_modified_accounts(ethdb::TransactionDatabase& tx_database, uint64_t start_block_number, uint64_t end_block_number) {
+Task<std::set<evmc::address>> get_modified_accounts(ethdb::TransactionDatabase& tx_database, BlockNum start_block_number, BlockNum end_block_number) {
```
Comment (canepat) [inline silkworm/silkrpc/commands/debug_api.cpp:628]: The same change should be applied also in `debug_api.hpp`

File: silkworm/silkrpc/commands/eth_api.cpp [hunk 2]
```diff
@@ -2200,8 +2200,8 @@ Task<void> EthereumRpcApi::handle_fee_history(const nlohmann::json& request, nlo
 
 Task<void> EthereumRpcApi::get_logs(
     ethdb::TransactionDatabase& tx_database,
-    std::uint64_t start,
-    std::uint64_t end,
+    BlockNum start,
+    BlockNum end,
```
Comment (canepat) [inline silkworm/silkrpc/commands/eth_api.cpp:2204]: The same change should be applied also in `eth_api.hpp`

File: silkworm/silkrpc/commands/ots_api.cpp [hunk 3]
```diff
@@ -844,14 +844,14 @@ Task<bool> OtsRpcApi::trace_blocks(
...
-Task<void> OtsRpcApi::search_trace_block(ethdb::Transaction& tx, evmc::address address, unsigned long index, uint64_t block_number, std::vector<TransactionsWithReceipts>& results) {
+Task<void> OtsRpcApi::search_trace_block(ethdb::Transaction& tx, evmc::address address, unsigned long index, BlockNum block_number, std::vector<TransactionsWithReceipts>& results) {
     TransactionsWithReceipts transactions_with_receipts;
     co_await trace_block(tx, block_number, address, transactions_with_receipts);
     results[index] = transactions_with_receipts;
     co_return;
 }
 
-Task<void> OtsRpcApi::trace_block(ethdb::Transaction& tx, uint64_t block_number, evmc::address search_addr, TransactionsWithReceipts& results) {
+Task<void> OtsRpcApi::trace_block(ethdb::Transaction& tx, BlockNum block_number, evmc::address search_addr, TransactionsWithReceipts& results) {
```
Comment (canepat) [inline silkworm/silkrpc/commands/ots_api.cpp:854]: The same change should be applied also in `ots_api.hpp`

File: silkworm/silkrpc/commands/ots_api.cpp [hunk 4]
```diff
@@ -966,7 +966,7 @@ ChunkProvider::ChunkProvider(silkworm::rpc::ethdb::Cursor* cursor, evmc::address
     first_seek_key_value_ = first_seek_key_value;
 }
 
-Task<ChunkLocatorResponse> ChunkLocator::get(uint64_t min_block) {
+Task<ChunkLocatorResponse> ChunkLocator::get(BlockNum min_block) {
```
Comment (canepat) [inline silkworm/silkrpc/commands/ots_api.cpp:969]: The same change should be applied also in `ots_api.hpp`

File: silkworm/silkrpc/commands/ots_api.cpp [hunk 5]
```diff
@@ -1043,7 +1043,7 @@ Task<BlockProviderResponse> ForwardBlockProvider::get() {
         }
     }
 
-    uint64_t next_block = next();
+    BlockNum next_block = next();
```
Comment (canepat) [inline silkworm/silkrpc/commands/ots_api.cpp:1046]: The same change should be applied also in `ForwardBlockProvider::next` method both in `ots_api.hpp` and `ots_api.cpp`

File: silkworm/silkrpc/core/account_dumper.cpp [hunk 6]
```diff
@@ -115,7 +115,7 @@ Task<void> AccountDumper::load_accounts(ethdb::TransactionDatabase& tx_database,
     co_return;
 }
 
-Task<void> AccountDumper::load_storage(uint64_t block_number, DumpAccounts& dump_accounts) {
+Task<void> AccountDumper::load_storage(BlockNum block_number, DumpAccounts& dump_accounts) {
```
Comment (canepat) [inline silkworm/silkrpc/core/account_dumper.cpp:118]: The same change should be applied also in `account_dumper.hpp`

File: silkworm/silkrpc/core/account_walker.cpp [hunk 7]
```diff
@@ -25,7 +25,7 @@
 
 namespace silkworm::rpc {
 
-Task<void> AccountWalker::walk_of_accounts(uint64_t block_number, const evmc::address& start_address, Collector& collector) {
+Task<void> AccountWalker::walk_of_accounts(BlockNum block_number, const evmc::address& start_address, Collector& collector) {
```
Comment (canepat) [inline silkworm/silkrpc/core/account_walker.cpp:28]: The same change should be applied also in `account_walker.hpp`

File: silkworm/silkrpc/core/account_walker.cpp [hunk 8]
```diff
@@ -110,13 +110,13 @@ Task<ethdb::SplittedKeyValue> AccountWalker::next(ethdb::SplitCursor& cursor, ui
     co_return skv;
 }
 
-Task<ethdb::SplittedKeyValue> AccountWalker::seek(ethdb::SplitCursor& cursor, uint64_t number) {
+Task<ethdb::SplittedKeyValue> AccountWalker::seek(ethdb::SplitCursor& cursor, BlockNum number) {
```
Comment (canepat) [inline silkworm/silkrpc/core/account_walker.cpp:113]: The same change should be applied also in `account_walker.hpp`

File: silkworm/silkrpc/core/block_reader.cpp [hunk 9]
```diff
@@ -73,7 +73,7 @@ Task<void> BlockReader::read_balance_changes(BlockCache& cache, const BlockNumbe
     co_return;
 }
 
-Task<void> BlockReader::load_addresses(uint64_t block_number, BalanceChanges& balance_changes) const {
+Task<void> BlockReader::load_addresses(BlockNum block_number, BalanceChanges& balance_changes) const {
```
Comment (canepat) [inline silkworm/silkrpc/core/block_reader.cpp:76]: The same change should be applied also in `block_reader.hpp`

File: silkworm/silkrpc/core/blocks.cpp [hunk 10]
```diff
@@ -61,8 +61,8 @@ Task<uint64_t> get_block_number_by_tag(const std::string& block_id, const rawdb:
     co_return block_number;
 }
 
-Task<std::pair<uint64_t, bool>> get_block_number(const std::string& block_id, const rawdb::DatabaseReader& reader, bool latest_required) {
-    uint64_t block_number;
+Task<std::pair<BlockNum, bool>> get_block_number(const std::string& block_id, const rawdb::DatabaseReader& reader, bool latest_required) {
```
Comment (canepat) [inline silkworm/silkrpc/core/blocks.cpp:64]: The same change should be applied also in `blocks.hpp`

File: silkworm/silkrpc/core/blocks.cpp [hunk 11]
```diff
@@ -93,12 +93,12 @@ Task<std::pair<uint64_t, bool>> get_block_number(const std::string& block_id, co
     co_return std::make_pair(block_number, is_latest_block);
 }
 
-Task<uint64_t> get_block_number(const std::string& block_id, const rawdb::DatabaseReader& reader) {
+Task<BlockNum> get_block_number(const std::string& block_id, const rawdb::DatabaseReader& reader) {
```
Comment (canepat) [inline silkworm/silkrpc/core/blocks.cpp:96]: The same change here and in the rest of functions in the module should be applied also in `blocks.hpp`.

File: silkworm/silkrpc/core/fee_history_oracle.cpp [hunk 12]
```diff
@@ -52,7 +52,7 @@ void to_json(nlohmann::json& json, const FeeHistory& fh) {
     }
 }
 
-Task<FeeHistory> FeeHistoryOracle::fee_history(uint64_t newest_block, uint64_t block_count, const std::vector<std::int8_t>& reward_percentile) {
+Task<FeeHistory> FeeHistoryOracle::fee_history(BlockNum newest_block, BlockNum block_count, const std::vector<std::int8_t>& reward_percentile) {
```
Comment (canepat) [inline silkworm/silkrpc/core/fee_history_oracle.cpp:55]: The same change should be applied also in `fee_history_oracle.hpp`

File: silkworm/silkrpc/core/fee_history_oracle.cpp [hunk 13]
```diff
@@ -119,7 +119,7 @@ Task<FeeHistory> FeeHistoryOracle::fee_history(uint64_t newest_block, uint64_t b
     co_return fee_history;
 }
 
-Task<BlockRange> FeeHistoryOracle::resolve_block_range(uint64_t last_block, uint64_t block_count, uint64_t max_history) {
+Task<BlockRange> FeeHistoryOracle::resolve_block_range(BlockNum last_block, uint64_t block_count, uint64_t max_history) {
```
Comment (canepat) [inline silkworm/silkrpc/core/fee_history_oracle.cpp:122]: The same change should be applied also in `fee_history_oracle.hpp`

File: silkworm/silkrpc/core/gas_price_oracle.cpp [hunk 14]
```diff
@@ -32,7 +32,7 @@ struct PriceComparator {
     }
 };
 
-Task<intx::uint256> GasPriceOracle::suggested_price(uint64_t block_number) {
+Task<intx::uint256> GasPriceOracle::suggested_price(BlockNum block_number) {
```
Comment (canepat) [inline silkworm/silkrpc/core/gas_price_oracle.cpp:35]: The same change should be applied also in `gas_price_oracle.hpp`

File: silkworm/silkrpc/core/gas_price_oracle.cpp [hunk 15]
```diff
@@ -64,7 +64,7 @@ Task<intx::uint256> GasPriceOracle::suggested_price(uint64_t block_number) {
     co_return price;
 }
 
-Task<void> GasPriceOracle::load_block_prices(uint64_t block_number, uint64_t limit, std::vector<intx::uint256>& tx_prices) {
+Task<void> GasPriceOracle::load_block_prices(BlockNum block_number, uint64_t limit, std::vector<intx::uint256>& tx_prices) {
```
Comment (canepat) [inline silkworm/silkrpc/core/gas_price_oracle.cpp:67]: The same change should be applied also in `gas_price_oracle.hpp`

File: silkworm/silkrpc/core/override_state.cpp [hunk 16]
```diff
@@ -84,26 +84,26 @@ evmc::bytes32 OverrideState::read_storage(const evmc::address& address, uint64_t
     return storage_value;
 }
 
-std::optional<silkworm::BlockHeader> OverrideState::read_header(uint64_t block_number, const evmc::bytes32& block_hash) const noexcept {
+std::optional<silkworm::BlockHeader> OverrideState::read_header(BlockNum block_number, const evmc::bytes32& block_hash) const noexcept {
```
Comment (canepat) [inline silkworm/silkrpc/core/override_state.cpp:87]: The same change here and in the rest of functions in the module should be applied also in `override_state.hpp`.

File: silkworm/silkrpc/core/remote_state.cpp [hunk 17]
```diff
@@ -54,28 +54,28 @@ Task<uint64_t> AsyncRemoteState::previous_incarnation(const evmc::address& /*add
     co_return 0;
 }
 
-Task<std::optional<silkworm::BlockHeader>> AsyncRemoteState::read_header(uint64_t block_number, const evmc::bytes32& block_hash) const noexcept {
+Task<std::optional<silkworm::BlockHeader>> AsyncRemoteState::read_header(BlockNum block_number, const evmc::bytes32& block_hash) const noexcept {
```
Comment (canepat) [inline silkworm/silkrpc/core/remote_state.cpp:57]: The same change here and in the rest of functions in the module should be applied also in `remote_state.hpp`.

File: silkworm/silkrpc/core/state_reader.cpp [hunk 18]
```diff
@@ -28,7 +28,7 @@
 
 namespace silkworm::rpc {
 
-Task<std::optional<silkworm::Account>> StateReader::read_account(const evmc::address& address, uint64_t block_number) const {
+Task<std::optional<silkworm::Account>> StateReader::read_account(const evmc::address& address, BlockNum block_number) const {
```
Comment (canepat) [inline silkworm/silkrpc/core/state_reader.cpp:31]: The same change here and in the rest of functions in the module should be applied also in `state_reader.hpp`.

File: silkworm/silkrpc/core/storage_walker.cpp [hunk 19]
```diff
@@ -89,7 +89,7 @@ Task<ethdb::SplittedKeyValue> next(ethdb::SplitCursor& cursor, uint64_t number,
 }
 
 Task<void> StorageWalker::walk_of_storages(
-    uint64_t block_number,
+    BlockNum block_number,
```
Comment (canepat) [inline silkworm/silkrpc/core/storage_walker.cpp:92]: The same change should be applied also in `storage_walker.hpp`.

**Review by canepat — CHANGES_REQUESTED** [review]

File: silkworm/silkrpc/commands/ots_api.cpp [hunk 20]
```diff
@@ -844,14 +844,14 @@ Task<bool> OtsRpcApi::trace_blocks(
     co_return has_more;
 }
 
-Task<void> OtsRpcApi::search_trace_block(ethdb::Transaction& tx, evmc::address address, unsigned long index, uint64_t block_number, std::vector<TransactionsWithReceipts>& results) {
+Task<void> OtsRpcApi::search_trace_block(ethdb::Transaction& tx, evmc::address address, unsigned long index, BlockNum block_number, std::vector<TransactionsWithReceipts>& results) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/commands/ots_api.cpp:847]: Ok

File: silkworm/silkrpc/commands/ots_api.cpp (hunk 3 above)
Comment (lupin012) [reply silkworm/silkrpc/commands/ots_api.cpp:854]: OK

File: silkworm/silkrpc/commands/ots_api.cpp (hunk 4 above)
Comment (lupin012) [reply silkworm/silkrpc/commands/ots_api.cpp:969]: ok

File: silkworm/silkrpc/commands/ots_api.cpp (hunk 5 above)
Comment (lupin012) [reply silkworm/silkrpc/commands/ots_api.cpp:1046]: ok

File: silkworm/silkrpc/core/account_dumper.cpp (hunk 6 above)
Comment (lupin012) [reply silkworm/silkrpc/core/account_dumper.cpp:118]: ok

File: silkworm/silkrpc/core/account_walker.cpp [hunk 21]
```diff
@@ -95,7 +95,7 @@ Task<KeyValue> AccountWalker::seek(ethdb::Cursor& cursor, silkworm::ByteView key
     co_return kv;
 }
 
-Task<ethdb::SplittedKeyValue> AccountWalker::next(ethdb::SplitCursor& cursor, uint64_t number, uint64_t block, silkworm::Bytes addr) {
+Task<ethdb::SplittedKeyValue> AccountWalker::next(ethdb::SplitCursor& cursor, BlockNum number, BlockNum block, silkworm::Bytes addr) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/core/account_walker.cpp:98]: ok

File: silkworm/silkrpc/core/account_walker.cpp (hunk 8 above)
Comment (lupin012) [reply silkworm/silkrpc/core/account_walker.cpp:113]: ok

File: silkworm/silkrpc/core/block_reader.cpp (hunk 9 above)
Comment (lupin012) [reply silkworm/silkrpc/core/block_reader.cpp:76]: ok

File: silkworm/silkrpc/core/blocks.cpp [hunk 22]
```diff
@@ -39,13 +39,13 @@ static Task<uint64_t> get_forkchoice_block_number(const rawdb::DatabaseReader& r
...
 
-Task<bool> is_latest_block_number(uint64_t block_number, const rawdb::DatabaseReader& db_reader) {
+Task<bool> is_latest_block_number(BlockNum block_number, const rawdb::DatabaseReader& db_reader) {
     const auto last_executed_block_number = co_await get_latest_executed_block_number(db_reader);
     co_return last_executed_block_number == block_number;
 }
 
-Task<uint64_t> get_block_number_by_tag(const std::string& block_id, const rawdb::DatabaseReader& reader) {
-    uint64_t block_number;
+Task<BlockNum> get_block_number_by_tag(const std::string& block_id, const rawdb::DatabaseReader& reader) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/core/blocks.cpp:47]: ok

File: silkworm/silkrpc/core/blocks.cpp (hunk 10 above)
Comment (lupin012) [reply silkworm/silkrpc/core/blocks.cpp:64]: ok

File: silkworm/silkrpc/core/blocks.cpp (hunk 11 above)
Comment (lupin012) [reply silkworm/silkrpc/core/blocks.cpp:96]: ok

File: silkworm/silkrpc/core/evm_debug.cpp [hunk 23]
```diff
@@ -271,7 +271,7 @@ void DebugTracer::write_log(const DebugLog& log) {
     stream_.write_json(json);
 }
 
-Task<void> DebugExecutor::trace_block(json::Stream& stream, const ChainStorage& storage, std::uint64_t block_number) {
+Task<void> DebugExecutor::trace_block(json::Stream& stream, const ChainStorage& storage, BlockNum block_number) {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/core/evm_debug.cpp:274]: ok

File: silkworm/silkrpc/core/fee_history_oracle.cpp (hunk 12 above)
Comment (lupin012) [reply silkworm/silkrpc/core/fee_history_oracle.cpp:55]: ok

File: silkworm/silkrpc/core/fee_history_oracle.cpp (hunk 13 above)
Comment (lupin012) [reply silkworm/silkrpc/core/fee_history_oracle.cpp:122]: ok

File: silkworm/silkrpc/core/gas_price_oracle.cpp (hunk 14 above)
Comment (lupin012) [reply silkworm/silkrpc/core/gas_price_oracle.cpp:35]: ok

File: silkworm/silkrpc/core/gas_price_oracle.cpp (hunk 15 above)
Comment (lupin012) [reply silkworm/silkrpc/core/gas_price_oracle.cpp:67]: ok

File: silkworm/silkrpc/core/local_state.cpp [hunk 24]
```diff
@@ -45,28 +45,28 @@ uint64_t LocalState::previous_incarnation(const evmc::address& /*address*/) cons
     return 0;
 }
 
-std::optional<silkworm::BlockHeader> LocalState::read_header(uint64_t block_number, const evmc::bytes32& block_hash) const noexcept {
+std::optional<silkworm::BlockHeader> LocalState::read_header(BlockNum block_number, const evmc::bytes32& block_hash) const noexcept {
```
[the remark this answers is not in the document]
Comment (lupin012) [reply silkworm/silkrpc/core/local_state.cpp:48]: ok

File: silkworm/silkrpc/core/override_state.cpp (hunk 16 above)
Comment (lupin012) [reply silkworm/silkrpc/core/override_state.cpp:87]: ok

File: silkworm/silkrpc/core/remote_state.cpp (hunk 17 above)
Comment (lupin012) [reply silkworm/silkrpc/core/remote_state.cpp:57]: ok

File: silkworm/silkrpc/core/state_reader.cpp (hunk 18 above)
Comment (lupin012) [reply silkworm/silkrpc/core/state_reader.cpp:31]: ok

File: silkworm/silkrpc/core/storage_walker.cpp (hunk 19 above)
Comment (lupin012) [reply silkworm/silkrpc/core/storage_walker.cpp:92]: ok

**Review by canepat — APPROVED** [review]

**Pull request merged**